keystore/src/connection/platform/wasm/migrations.rs

```diff
@@ -0,0 +1,406 @@
+use keystore_v_1_0_0::entities::{
+    E2eiAcmeCA as E2eiAcmeCAV1_0_0, E2eiCrl as E2eiCrlV1_0_0, E2eiEnrollment as E2eiEnrollmentV1_0_0,
+    E2eiIntermediateCert as E2eiIntermediateCertV1_0_0, E2eiRefreshToken as E2eiRefreshTokenV1_0_0,
+    Entity as EntityV1_0_0, MlsCredential as MlsCredentialV1_0_0, MlsEncryptionKeyPair as MlsEncryptionKeyPairV1_0_0,
+    MlsEpochEncryptionKeyPair as MlsEpochEncryptionKeyPairV1_0_0, MlsHpkePrivateKey as MlsHpkePrivateKeyV1_0_0,
+    MlsKeyPackage as MlsKeyPackageV1_0_0, MlsPendingMessage as MlsPendingMessageV1_0_0,
+    MlsPskBundle as MlsPskBundleV1_0_0, MlsSignatureKeyPair as MlsSignatureKeyPairV1_0_0,
+    PersistedMlsGroup as PersistedMlsGroupV1_0_0, PersistedMlsPendingGroup as PersistedMlsPendingGroupV1_0_0,
+    ProteusIdentity as ProteusIdentityV1_0_0, ProteusPrekey as ProteusPrekeyV1_0_0,
+    ProteusSession as ProteusSessionV1_0_0, UniqueEntity as UniqueEntityV1_0_0,
+};
+use keystore_v_1_0_0::CryptoKeystoreError as CryptoKeystoreErrorV1_0_0;
+
+const fn db_version_number(counter: u32) -> u32 {
+    // This is how version numbers were calculated when the DB version was tied to core crypto versions
+    // 10_000_000 for a major version, 1_000 for a patch version. I.e., the number for v1.0.2 was
+    const VERSION_1_0_2: u32 = 10_000_000 + 2_000;
+    // From post v1.0.2, we will just increment whenever we need a DB migration.
+    VERSION_1_0_2 + counter
+}
+
+const DB_VERSION_0: u32 = db_version_number(0);
+
+/// Open an existing idb database with the given name and key, and migrate it if needed.
+pub(crate) async fn open_and_migrate(name: &str, key: &str) -> CryptoKeystoreResult<Database> {
+    /// Increment when adding a new migration.
+    const TARGET_VERSION: u32 = db_version_number(1);
+    let factory = Factory::new()?;
+
+    let open_existing = factory.open(name, None)?;
+    let existing_db = open_existing.await?;
+    let mut version = existing_db.version()?;
+    if version == TARGET_VERSION {
+        // Migration is not needed, just return existing db
+        Ok(existing_db)
+    } else {
+        // Migration is needed
+        existing_db.close();
+
+        while version < TARGET_VERSION {
+            version = do_migration_step(version, name, key).await?;
+        }
+
+        let open_request = factory.open(name, Some(TARGET_VERSION))?;
+        open_request.await.map_err(Into::into)
+    }
+}
+
+/// This is called from a while loop. The `from` argument represents the version the migration is performed from.
+/// The function will return the version number of the DB resulting from the migration.
+///
+/// To add a new migration, add a new match arm below the latest one.
+/// It must match on the version it migrates from, and call a function that performs the migration workload,
+///     which returns the version it migrates to, which is the same value as TARGET_VERSION in the function above at the time the migration is added.
```

Comment:
Nitpick: this line is unnecessarily indented and also a bit long, please consider breaking it.